src/handlers/VehicleTestController.ts

```diff
@@ -206,25 +268,213 @@ export class VehicleTestController implements IVehicleTestController {
     return payload;
   }
 
-  private static calculateAnniversaryDate(payload: models.ITestResultPayload) {
-    const { vehicleType } = payload;
-    payload.testTypes.forEach((testType: models.TestType) => {
-      const { testExpiryDate } = testType;
-      if (!testExpiryDate) {
-        return;
+  private async mapOldTestResultToNew(
+    systemNumber: string,
+    payload: models.ITestResult,
+    msUserDetails: models.IMsUserDetails
+  ) {
+    const testResults = await this.dataProvider.getBySystemNumber(systemNumber);
+    const oldTestResult = VehicleTestController.getTestResultToArchive(
+      testResults,
+      payload.testResultId
+    );
+    const newTestResult = cloneDeep(oldTestResult);
+    oldTestResult.testVersion = enums.TEST_VERSION.ARCHIVED;
+    newTestResult.testVersion = enums.TEST_VERSION.CURRENT;
+    mergeWith(newTestResult, payload, utils.MappingUtil.arrayCustomizer);
+    newTestResult.testTypes = await this.generateNewTestCode(
+      oldTestResult,
+      newTestResult
+    );
+    await this.checkTestTypeStartAndEndTimestamp(oldTestResult, newTestResult);
+    utils.MappingUtil.cleanDefectsArrayForSpecialistTests(newTestResult);
+    utils.MappingUtil.setAuditDetails(
+      newTestResult,
+      oldTestResult,
+      msUserDetails
+    );
+    if (!newTestResult.testHistory) {
+      newTestResult.testHistory = [oldTestResult];
+    } else {
+      delete oldTestResult.testHistory;
+      newTestResult.testHistory.push(oldTestResult);
+    }
+    return newTestResult;
+  }
+
+  private async generateNewTestCode(
+    oldTestResult: models.ITestResult,
+    newTestResult: models.ITestResult
+  ) {
+    const {
+      vehicleType,
+      vehicleSize,
+      vehicleConfiguration,
+      euVehicleCategory,
+      testTypes,
+    } = newTestResult;
+    if (
+      !VehicleTestController.shouldGenerateNewTestCode(
+        oldTestResult,
+        newTestResult
+      )
+    ) {
+      return testTypes;
+    }
+    const vehicleSubclass =
+      newTestResult.vehicleSubclass && newTestResult.vehicleSubclass.length
+        ? newTestResult.vehicleSubclass[0]
+        : undefined;
```

Comment:
i will need a catch up on that one to understand the specs better please

Reply:
This is the same logic as originally on develop. It is not modified to avoid failures.